app/services/report_importer.rb

```diff
@@ -100,9 +100,12 @@ def notify_on_report_error(mail_error_state)
       end
 
       owners = host.owner.present? ? host.owner.recipients_for(:puppet_error_state) : []
+      users = MailNotification::PuppetError.all_hosts.flat_map(& :users)
+      users.select { |user| Host.authorized_as(user, :view_hosts).find(host.id).present? }
```

Comment:
The `users` contains all users who subscribed to `all hosts`, and before sending an email, need to authorize if they can view the reported host.
I don't see a way to move this logic to the `PuppetError` class.